### PR TITLE
metrics: record time to update gc info as a per timeline metric

### PR DESCRIPTION
**Pull request by koivunej**
We know that updating gc info can take a very long time from [recent incident], and holding `Tenant::gc_cs` affects many per-tenant operations in the system. We need a direct way to observe the time it takes. The solution is to add metrics so that we know when this happens:
- 2 new per-timeline metric
- 1 new global histogram

Verified that the buckets are okay-ish in [dashboard]. In our current state, we will see a lot more of `Inf,` but that is probably okay; at least we can learn which timelines are having issues.

Can we afford to add these metrics? A bit unclear, see [another dashboard] with top pageserver `/metrics` response sizes.

[dashboard]: https://neonprod.grafana.net/d/b7a5a5e2-1276-4bb0-9e3a-b4528adb6eb6/storage-operations-histograms-in-prod?orgId=1&var-datasource=ZNX49CDVz&var-instance=All&var-operation=All&from=now-7d&to=now

[another dashboard]: https://neonprod.grafana.net/d/MQx4SN-Vk/metric-sizes-on-prod-and-some-correlations?orgId=1

[recent incident]: https://neondb.slack.com/archives/C06UEMLK7FE/p1713817696580119?thread_ts=1713468604.508969&cid=C06UEMLK7FE

### CODE REVIEW

File: pageserver/src/tenant.rs [hunk 1]
```diff
@@ -3399,7 +3399,11 @@ impl Tenant {
         // is in progress (which is not a common case).
         //
         // See more for on the issue #2748 condenced out of the initial PR review.
-        let mut shared_cache = self.cached_logical_sizes.lock().await;
+        let mut shared_cache = tokio::select! {
+            locked = self.cached_logical_sizes.lock() => locked,
+            _ = cancel.cancelled() => anyhow::bail!("cancelled"),
+            _ = self.cancel.cancelled() => anyhow::bail!("tenant is shutting down"),
+        };
```
Comment (problame) [inline pageserver/src/tenant.rs:3406]: I like this change but the PR description doesn't mention it. Would be cleaner to have separate PRs for behavioral change and "just metrics".

File: pageserver/src/tenant/size.rs [hunk 2]
```diff
@@ -118,6 +118,9 @@ pub(super) async fn gather_inputs(
     ctx: &RequestContext,
 ) -> anyhow::Result<ModelInputs> {
     // refresh is needed to update gc related pitr_cutoff and horizon_cutoff
+    //
+    // FIXME: if a single timeline is deleted while refresh gc info is ongoing, we will fail the
+    // whole computation. It does not make sense from the billing perspective.
```
Comment (problame) [inline pageserver/src/tenant/size.rs:123]: Would like to see a link to the bug (in Selected category) in this comment.

File: pageserver/src/tenant/timeline.rs [hunk 3]
```diff
@@ -4149,6 +4149,12 @@ impl Timeline {
         cancel: &CancellationToken,
         ctx: &RequestContext,
     ) -> anyhow::Result<()> {
+        let _timer = self
+            .metrics
+            .update_gc_info_histo
+            .start_timer()
+            .record_on_drop();
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:4156]: I think `record_on_drop` is a mistake here. If there's a broken tenant in the future that bails out early, it will skew the histogram.

ALso, I expect we'll have a multimodal histogram in the regular case anyways.
One peak near zero with pitr_interval=0 because that caode path is super cheap.
Then one peak at pitr_interval=1d (free tier).
And another peak at pitr_interval=7d (default for paid plans).

Why different peaks? Because `find_lsn_for_timestamp` runtime depends on gc_horizon, which is determined primarily by pitr_interval.

**Review by problame — APPROVED** [review]
No strong objections, some comments.
And I don't have opinions on whether we can afford the additional metrics.

File: pageserver/src/tenant.rs (hunk 1 above)
Comment (koivunej) [reply pageserver/src/tenant.rs:3406]: I am not exactly sure if this fixes anything, I just added it in a whim because I was going for a hotfix and it cannot hurt.

This is likely unnecessary because everything in synthetic size calculation is fast OR cancellation-sensitive already (logical sizes, update gc info).

File: pageserver/src/tenant/timeline.rs (hunk 3 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:4156]: Agreed, it will have multiple peaks. However, I thought the per-timeline values (or the per-tenant sum) would still be useful to catch durations in which branch creation will be slow.

File: pageserver/src/tenant/size.rs (hunk 2 above)
Comment (koivunej) [reply pageserver/src/tenant/size.rs:123]: That's just a guess, assuming refresh_gc_info is still a problem. Could be it's not.

**Comment by koivunej** [comment]
We discussed this more last week. The guess is that we should have room for this metric, so merging.
